### DISCUSSION TITLE
State of Ontology Manager

### DISCUSSION THREAD

**Discussion by ktk** [zazuko/ontology-manager#75] [General]
Last week I met with some people that are interested in using Ontology Manager and I gave some feedback about where I see it and where not. Might make sense to share this publicly so we can see what might happen from here.

The TL;DR story of this software is that we developed that for a client based on their requirements. The goal was to mimic the schema.org approach: A community driven schema that gets extended when there is a clear need and where people have a discussion before a new class or property is added. This is unlike "standard" schema or ontologies, that are often created in tools like Protégé by a small group of people.

The problem with the schema.org approach is that it is still techie. So the client asked us to create a "frontend" that domain specialists could use instead. After a bunch of sprints we created what I would call Ontology Manager 1.0, what you can find here as Open Source Software.

There 1.0 version provides these features:
* A web fronted where the ontology can be explored,
* and where new classes and properties can be requested
* A strong focus on justification. The first thing you have to fill out is a description on *why* you think this class is needed.
* A way to have a discussion. New people might look for something that was rejected before so we want to be able to point them to the previous discussion so they understand why it won't happen.
* A way to moderate & reject proposals. This is the admin-view in the Ontology Manager. Someone has to accept or reject something before it is merged into the "main" branch.
* A way to deprecate things.
* A way to link to similar concepts elsewhere.
* The main branch is always versioned in Git so we can see who requested what. This currently relies on the GitHub API so it can only be deployed on Github right now.

For reasons I still don't fully understand we have not seen people using Ontology Manager, at least not that we noticed. Last week I found an old but very relevant post about [The Rule of Three](https://blog.codinghorror.com/rule-of-three/) which basically says:

> There are two "rules of three" in software reuse:
>
>  * It is three times as difficult to build reusable components as single use components, and
>  * a reusable component should be tried out in three different applications before it will be sufficiently general to accept into a reuse library.

While the original quote talks about libraries, Jeff says that's their rule for "applications" as well. And in that regard, to the best of my knowledge, Ontology Manager failed so far. The client that payed it did not use it (their division got dismantled once we were done so we built it for no one) and other instances we run did not see much uptake.

About the reasons I'm not sure. Maybe these groups simply didn't manage to build a community. I'm convinced that without community, the software is pretty much pointless. But it could also be that the UX is completely wrong or some other, yet unknown reason.

I have multiple ideas of what I would like to do with it, number one idea is probably to add support for SHACL shapes. SHACL was specified pretty much around the time we did it but back then I did not understand yet how important SHACL will become for us. So if I would have customer interest (aka money) to work on it, I would probably think about SHACL integration, making it part of what people have to describe. I would also invest time into UX testing to see how people work with it or why they don't work with it. And I would spend time refactoring quite some of the backend stuff. Right now it still depends on PostgreSQL for some reasons. I would kick that out and probably implement the discussions using Activity Pub protocol. That would be way more linked data.

But nothing makes sense unless we are sure that it's a problem worth solving and for that we need to show that we have at least three different success stories.

What's your take on this?

**Comment by ktk** [comment]
I [tweeted](https://twitter.com/linkedktk/status/1620060836537696256) about it and got a [question](https://twitter.com/FanLi_RnD/status/1620063325316071426) which makes sense to answer here as well: What was the use-case?

I think I can talk about that, the client was in the "shipping" industry. As in a large industrial company that ships a lot of containers world wide and is completely annoyed by the analog process involved in it. Apparently shipping is (probably until now) very much analog, as in people fill out the same things many, many times.

So this was part of a consortium of other large companies that want to address that. As you can imagine this is most likely a *very* hard problem to solve due to all the involved parties and their interest. Some came to the conclusion that this can only be solved using Linked Data (I agree obviously) but this would involve defining a "schema" or "ontology" to have the interfaces clearly semantically specified. And then it became clear that no single entity can propose such a schema, as others would refuse it for the simple fact that it was not them proposing it. Yes, that level of discussions...

So for a while there was a now defunct instance of Ontology Manager running under a neutral domain that was supposed to be driven by all the different participants from these different companies and organizations. But as I mentioned, the main driver got dismantled before there was anything like a community and it fell apart pretty quickly after that.

**Comment by tpluscode** [comment]
When I'd tried it in the past I found it rather complicated to run. Both in dev environment and to run a container. Maybe I was less skilled with docker at that point but could be a factor.

And to play the advocate too, it is quite possible that the issue was lacking documentation or a "batteries-included" starter pack and not the tool itself

**Reply by cristianvasquez** [reply]
The same happened to me; only a few people have the time to try out something that takes more than 5 minutes to run. 

It isn't straightforward; today it's a bit overcomplicated.

**Reply by ktk** [reply]
Docker might be an option, the setup right now is relatively easy to deploy in environments where there is hosted Postgres for example.

Getting rid of the Postgres dependency would surely be better but lots of work. The Github setup itself is also not straight forward.

**Comment by burakemir** [comment]
I don't have much context but I hope these remarks can still be useful:

* domain model, schema and ontology are things that developers encounter on the path from requirements to implementation. Only developers/experts understand the benefits of machine-readable representation of "real world" concepts and very often they elaborate in the context of one particular application. Users have no experience of these benefits and may be inclined to think that they have to deal with all the jargon and in the end it only makes the developers' lives easier.
* there are no established quality criteria that would help assess how "complete", accurate, precise or redundant a domain model is. Developers of an application will focus on a single application but if you target a broad range of applications than that focus does not help
* in terms of justifying/deprecating etc evolving is a deeper challenge that will even baffle experts (with different perspectives). Beyond changes to schema/domain model/ontology there usually is a transition process involves where old uses have to be updated to the new. This process is beyond any single tool. For the non expert it is additionally confusing since the evolution of meanings (and other expectations, non-functional requirements) happens invisibly and on a larger time scale. It should not be "at the click of a button" or at least the tool should connect actions with the *consequences* of these actions which may be outside the tool. Measuring progress of such transitions is often possible but keeping the overview can be tough as each transition is different

**Comment by hobofan** [comment]
Very interesting question. I'll try to share my thoughts from the perspective of someone who was generally interested in adopting "some ontology editor" as part of our solutions with both/either the [Field33 SaaS product](https://field33.com/) or the [Plow ontology package manager](https://plow.pm/) and stumbled upon ZOM in the process. 

# General

- It's a "full app", including a stateful database, which would make adopting it a **huge** decision. Especially since it's user-facing, where data loss may mean angry customers/community.
- It has AGPL as a license. Especially for an app like this, where you would potentially want to swap out some parts and adjust them to make them work well with the rest of your stack, that's quite the dealbreaker. While we don't have a legal department that expressively forbids that license, conforming to the license would mean a very different development workflow to our usual one, which is a big downside.
- While a bit sad, I think with a reactive frontend, the stack may be "too modern" for quite some teams/companies in the space to adopt, where frontend expertise seems to be rare in general

# ZOM + Plow

As Plow is also a "manager" of ontologies, one would think that it would be in "competition" with ZOM here, which I don't think needs to be the case. 

Right now, ontologies on Plow are mostly created via one of three ways:
- Mirroring of external ontologies (where we add some Plow-related metadata, but otherwise leave the ontology as-is)
- Manually edited via Protege (which we are not so happy with)
- Machinally created based on e.g. CSV data + some scripts

As Plow isn't exclusive to "one true way" of ontology editing, I think for some communities managing packages on there something like ZOM might actually be a good fit!

However in general, there seems to be a bit of a misalignment in philosophies here, which I think leads to some friction.

> schema.org approach

> This is unlike "standard" schema or ontologies, that are often created in tools like Protégé by a small group of people.

I don't think that the schema.org approach is that desirable, and I don't think that the "standard" approach is the worst either (but it has some big warts in its current form). What I think is a good fit for an open ecosystem is consensus-by-usage (small group publishes, and then it's either adopted or not) rather than consensus-by-committee (schema.org and similar).

However, both approaches are currently being practiced with waaaay too big ontologies. Plow instead tries to put an emphasis on modularity and re-use with smaller ontologies, where you can depend on only the ontologies you need, rather than depending on a few big monolithic ones, where half of the concepts may be misaligned for you.

So back to integrating ZOM and Plow:
- As far as I can tell it's 1 ontology per server, which again incentivizes an ecosystem with a few big ontologies. I think for it to be a better fit, it would at least need to be possible to manage multiple packages from the same Plow namespace
- ZOM seems to be purely git versioned, while for Plow packages, there are dedicated SemVer-versioned releases. That doesn't seem like too big of a gap, where ZOM can represent the "in development" version of an ontology, and the connection to Plow releases could be recorded.
- To be a better fit we would probably like to tweak the external reference resolution to be aware of Plow dependencies.

So overall, I think for ZOM + Plow, our biggest blockers to adopting it / making it easier for Plow-related communities, would still be the "General" concerns listed above. Beyond that, I think it depends on how amendable ZOM would be to support a multi-ontology model.

# ZOM + Field 33 product

In the Field 33 product, each organization represents a separate tenant, and each of them can load separate ontologies via Plow. In addition to that, there is some (right now unversioned) ontology editing, which is right now mostly confined to OWL ABox constructs + some bespoke constructs that are custom to the platform and not (yet) standardized for external usage.

- Here we would integrate ZOM probably quite deeply into the product to get a good UX, meaning a lot of adjustments
- Would probably involve adding a new forge backend
- Vue is a bit of a mismatch to our React frontend, but I think that could be bridged
- Separate ontologies (1 per tenant) would need to be supported, otherwise hosting 1 ZOM instance per tenant will likely be too cost prohibitive

Again, the "General" concerns dominate of the specific ones here. With the amount of tweaks necessary to make it a good fit, we are also probably in a territory where we would consider creating our own ontology editor.

-----

Hope this provides some insights and that I didn't ramble too much! :)

**Reply by ktk** [reply]
Thanks & sorry for the long delay.



Let me comment on your topics.

# General

* Full app: Yes that is true. If we would have someone investing money I would surely rip out the Postgres dependency. There will always be a state as we want to capture both the ontology plus the discussions around that ontology. But I see a potential of moving the discussion to RDF as well so one could simply commit it to git as well. I think we might still need some active component for the webapp but data loss would be easy preventable that way.
* AGPL: It's our code, so if this is the show stopper, I'm happy to talk about another license that would be better. Our libraries are generally MIT or BSD. The reason we chose AGPL so far was because there is a notorious company in the Semweb space that takes existing OSS software, changes the style/CSS, and puts their name in their product offering without ever giving anything back. If you and others are more motivated to work with us, we can always change the license.
* Frontend: Yeah we had a phase where Vue was hip within Zazuko. That changed a bit meanwhile for multiple reasons.

# ZOM & Plow

Not sure I get plow. I had a look at plow.pm and I'm not sure I understand what it does or provides, docs looks very minimal so far? So can't comment much on that, maybe we would have to have a chat about the topic.

> What I think is a good fit for an open ecosystem is consensus-by-usage (small group publishes, and then it's either adopted or not) rather than consensus-by-committee (schema.org and similar).

I like that. 

> * As far as I can tell it's 1 ontology per server, which again  incentivizes an ecosystem with a few big ontologies. I think for it to  be a better fit, it would at least need to be possible to manage  multiple packages from the same Plow namespace

fully agree on that, we also have small ontologies which we publish. ZOM would have to be extended for that use case.

> * ZOM seems to be purely git versioned, while for Plow packages, there are dedicated SemVer-versioned releases. That doesn't seem like too big of a gap, where ZOM can represent the "in development" version of an  ontology, and the connection to Plow releases could be recorded.

absolutely, we discussed setups like this with customers.

> So overall, I think for ZOM + Plow, our biggest blockers  to adopting it / making it easier for Plow-related communities, would  still be the "General" concerns listed above. Beyond that, I think it  depends on how amendable ZOM would be to support a multi-ontology model.

multi-ontology model is surely a good idea. But it would definitely need some serious refactoring of the codebase.

# ZOM + Field 33 product

I have an idea of what you explain but as mentioned above I would need to understand Plow more.
